shiny/experimental/ui/_fill.py

```diff
@@ -254,85 +348,208 @@ def remove_all_fill(tag: TagT) -> TagT:
+    tag
+        a Tag object.
+
+
+    See Also
+    --------
+    * :func:`~shiny.experimental.ui.as_fill_carrier`
+    * :func:`~shiny.experimental.ui.as_fill_item`
+    * :func:`~shiny.experimental.ui.as_fillable_container`
+    * :func:`~shiny.experimental.ui.remove_all_fill`
+    * :func:`~shiny.experimental.ui.is_fill_carrier`
+    * :func:`~shiny.experimental.ui.is_fillable_container`
+    """
     # TODO-future; Handle widgets
     # # won't actually work until (htmltools#334) gets fixed
     # renders_to_tag_class(x, fill_item_class, ".html-widget")
 
-    return is_fill_layout(x, layout="fill")
+    return _is_fill_layout(x, layout="fill")
 
 
-def is_fill_layout(
-    x: TagChild | FillingLayout,
+def _is_fill_layout(
+    x: TagChild | IsFillingLayout,
     layout: Literal["fill", "fillable"],
     recurse: bool = True,
 ) -> bool:
-    if not isinstance(x, (Tag, Tagifiable, FillingLayout)):
+    if not isinstance(x, (Tag, Tagifiable, IsFillingLayout)):
         return False
 
-    # x: Tag | FillingLayout | Tagifiable
+    # x: Tag | IsFillingLayout | Tagifiable
 
     if layout == "fill":
         if isinstance(x, Tag):
             return x.has_class(fill_item_class)
-        if isinstance(x, FillingLayout):
+        if isinstance(x, IsFillingLayout):
             return x.is_fill_item()
 
     elif layout == "fillable":
         if isinstance(x, Tag):
             return x.has_class(fill_container_class)
-        if isinstance(x, FillingLayout):
+        if isinstance(x, IsFillingLayout):
             return x.is_fillable_container()
 
-    # x: Tagifiable and not (Tag or FillingLayout)
+    # x: Tagifiable and not (Tag or IsFillingLayout)
     raise TypeError(
-        f"`is_fill_layout(x=)` must be a `Tag` or implement the `FillingLayout` protocol methods TODO-barret expand on method names. Received object of type: `{type(x).__name__}`"
+        f"`_is_fill_layout(x=)` must be a `Tag` or implement the `IsFillingLayout` protocol methods TODO-barret expand on method names. Received object of type: `{type(x).__name__}`"
     )
 
 
 @runtime_checkable
-class FillingLayout(Protocol):
+class IsFillingLayout(Protocol):
     def is_fill_item(self) -> bool:
-        raise NotImplementedError()
+        ...
 
     def is_fillable_container(self) -> bool:
-        raise NotImplementedError()
+        ...
+
+
+T = TypeVar("T")
+
+
+@runtime_checkable
+class AsFillingLayout(Protocol):
+    def add_class(
+        self: T,
+        class_: str,
+        *,
+        # Currently Unused
+        css_selector: Optional[str] = None,
+    ) -> T:
+        ...
+
+    def add_style(
+        self: T,
+        style: str,
+        *,
+        # Currently Unused
+        css_selector: Optional[str] = None,
+    ) -> T:
+        ...
+
+    def as_fill_item(
+        self: T,
+    ) -> T:
+        ...
+
+    def as_fillable_container(
+        self: T,
+    ) -> T:
+        ...
+
+    def remove_all_fill(
+        self: T,
+    ) -> T:
+        ...
```

Comment:
Is this approach of using two different class values a fair approach to make this section generic for classes and not just Tag objects?